src/solvers/refinement/string_constraint.h

```diff
@@ -0,0 +1,181 @@
+	and not_contains_constraintt of the form:
+        $\forall x in [lb,ub[. p(x) => \exists y in [lb,ub[. s1[x+y] != s2[y]$.
+
+Author: Romain Brenguier, romain.brenguier@diffblue.com
+
+\*******************************************************************/
+
+#ifndef CPROVER_SOLVERS_REFINEMENT_STRING_CONSTRAINT_H
+#define CPROVER_SOLVERS_REFINEMENT_STRING_CONSTRAINT_H
+
+#include <langapi/language_ui.h>
+#include <solvers/refinement/bv_refinement.h>
+#include <solvers/refinement/refined_string_type.h>
+
+class string_constraintt: public exprt
+{
+public:
+  // String constraints are of the form
+  // forall univ_var in [lower_bound,upper_bound[. premise => body
+
+  inline const exprt &premise() const
+  {
+    return op0();
+  }
+
+  inline const exprt &body() const
+  {
+    return op1();
+  }
+
+  inline const symbol_exprt &univ_var() const
+  { return to_symbol_expr(op2()); }
+
+  inline const exprt &upper_bound() const
+  {
+    return op3();
+  }
+
+  inline const exprt &lower_bound() const
+  {
+    return operands()[4];
+  }
+
+
+ private:
+  string_constraintt();
+
+ public:
+  string_constraintt(
+    const symbol_exprt &univ,
+    const exprt &bound_inf,
+    const exprt &bound_sup,
+    const exprt &prem,
+    const exprt &body):
+    exprt(ID_string_constraint)
+  {
+    copy_to_operands(prem, body);
+    copy_to_operands(univ, bound_sup, bound_inf);
+  }
+
+  // Default bound inferior is 0
+  string_constraintt(
+    const symbol_exprt &univ,
+    const exprt &bound_sup,
+    const exprt &prem,
+    const exprt &body):
+    string_constraintt(
+      univ,
+      from_integer(0, univ.type()),
+      bound_sup,
+      prem,
+      body)
+  {}
+
+  // Default premise is true
+  string_constraintt(
+    const symbol_exprt &univ,
+    const exprt &bound_sup,
+    const exprt &body):
+    string_constraintt(univ, bound_sup, true_exprt(), body)
+  {}
+
+  inline exprt univ_within_bounds() const
+  {
+    return and_exprt
+      (binary_relation_exprt(lower_bound(), ID_le, univ_var()),
+       binary_relation_exprt(upper_bound(), ID_gt, univ_var()));
+  }
+};
+
+extern inline const string_constraintt &to_string_constraint(const exprt &expr)
+{
+  assert(expr.id()==ID_string_constraint && expr.operands().size()==5);
+  return static_cast<const string_constraintt &>(expr);
+}
+
+extern inline string_constraintt &to_string_constraint(exprt &expr)
+{
+  assert(expr.id()==ID_string_constraint && expr.operands().size()==5);
+  return static_cast<string_constraintt &>(expr);
+}
+
+class string_not_contains_constraintt: public exprt
+{
+public:
+  // string_not contains_constraintt are formula of the form:
+  // forall x in [lb,ub[. p(x) => exists y in [lb,ub[. s1[x+y] != s2[y]
+
+  string_not_contains_constraintt(
+    exprt univ_lower_bound,
+    exprt univ_bound_sup,
+    exprt premise,
+    exprt exists_bound_inf,
+    exprt exists_bound_sup,
+    exprt s0,
+    exprt s1)
+    :exprt(ID_string_not_contains_constraint)
```

Comment:
done https://github.com/diffblue/cbmc/pull/374/commits/4887c9d3c6025c7789963ff97a6b9cc7240a36d1 but I'm not sure what the indentation should be after that: should it the be the same as the constructor, as the argument or two more